### ISSUE TITLE
Migrate FunctionalGroupHierarchy and PreprocessReaction to C++ 

### ISSUE THREAD

**Issue by bp-kelley** [rdkit/rdkit#732]
Where should these go?

PreprocessReaction -> GraphMol/ChemReactions

FunctionalGroupHierarchy -> Substructs?

**Comment by greglandrum** [comment]
yes, PreprocessReaction -> ChemReactions

FunctionalGroupHierarchy is less obvious unless you are planning on adding a `HierarchicalMatcher` class or something like that.

**Comment by bp-kelley** [comment]
I think a Hierarchical Matcher is useful enough to include in Substructs.

**Comment by greglandrum** [comment]
Yeah, I can think of numerous uses for it. Maybe worth a discussion on Slack #development
